src/inmanta/data/__init__.py

```diff
@@ -2191,6 +2217,102 @@ async def delete_older_than(
         query = "DELETE FROM " + cls.table_name() + " WHERE completed <= $1::timestamp with time zone"
         await cls._execute_query(query, oldest_retained_date, connection=connection)
 
+    @classmethod
+    def _get_paging_item_count_query(
+        cls,
+        environment: uuid.UUID,
+        database_order: DatabaseOrder,
+        id_column_name: ColumnNameStr,
+        first_id: Optional[Union[uuid.UUID, str]] = None,
+        last_id: Optional[Union[uuid.UUID, str]] = None,
+        start: Optional[Any] = None,
+        end: Optional[Any] = None,
+        **query: Any,
+    ) -> Tuple[str, List[Any]]:
+        select_clause, values, common_filter_statements = cls._get_item_count_query_conditions(
+            database_order, id_column_name, first_id, last_id, start, end, **query
+        )
+        sql_query = f"""{select_clause}
+                    FROM {cls.table_name()}
+                    WHERE environment = ${len(values) + 1}
+                    """
+        values.append(cls._get_value(environment))
+        if len(common_filter_statements) > 0:
+            sql_query += cls._join_filter_statements_without_where(common_filter_statements)
+
+        return sql_query, values
+
+    @classmethod
+    async def get_compile_reports(
+        cls,
+        database_order: DatabaseOrder,
+        limit: int,
+        environment: uuid.UUID,
+        first_id: Optional[uuid.UUID] = None,
+        last_id: Optional[uuid.UUID] = None,
+        start: Optional[datetime.datetime] = None,
+        end: Optional[datetime.datetime] = None,
+        connection: Optional[asyncpg.connection.Connection] = None,
+        **query: Tuple[QueryType, object],
+    ) -> List[m.CompileReport]:
+        order_by_column = database_order.get_order_by_db_column_name()
```

Comment:
Stylistically, I think we should search for a way of improving this filtering/paging framework because

1. all pull requests have a lot of similar code, which suggest duplication
2. this code is quite hard to review, the abstractions of the methods and classes that are used are hard to follow. 

I'll try to give some more detailed comments with specific parts.


I wonder if a query builder class would help, that hides all the little mechanics of counting values and merging queries?

Reply:
I agree it's hard to review and improvements are welcome.

Reply:
Yes, I've been thinking about some refactoring for a while, but wasn't sure when to start.
I'll mark this as a draft while I start implementing the suggestions